### ISSUE TITLE
CarPlay support?

### ISSUE THREAD

**Issue by luke2m** [leinelissen/jellyfin-audio-player#44]
It would be great to have Apple CarPlay support added to this app. I’m not sure how hard this would be, but it seems to be possible with React Native.

**Comment by scgf** [comment]
Yes, I agree. I just setup Jellyfin and remote access and the app looks really beautiful. Playing music in my car I can access it on my CarPlay screen if I go to Now Playing, but there is no Car Play app for Jellyfin Audio Player.

It has been a year since @luke2m posted his comment. Is Car Play support being worked on?

**Comment by leinelissen** [comment]
Hey @scgf! Thanks for tuning in. I agree that getting CarPlay in would be a really cool feature and a nice challenge to work on. Unfortunately I maintain and extend this app in my spare time, and there are other issues and features that are currently higher on my priority list. I would love to see this happen, but this is not going to be picked up in the near future. Now if you are up for a challenge yourself, and if you have a knack for React Native stuff, I am more than happy to give a few pointers how you could come to an implementation. I would love to help out to get this in.

**Comment by scgf** [comment]
Thanks for your reply @leinelissen . Unfortunately I have no useful expertise with these things!

**Comment by jenkshields** [comment]
Kia ora! I've just started using Fintunes and it's great! I searched for this issue because I'd love to help improve the carplay support - I've got experience with react but not RN, but I'd love to hear those pointers if you're still interested!

**Comment by leinelissen** [comment]
Hi @jenkshields, thanks for using it from the other side of the planet! React Native is not that different from regular React, other than that the elements are a bit different (i.e. `<View />`s instead of `<div />`s) and the build process is a lot more complex and prone to breakage. 

You can try getting to a build first: firstly, `clone` the repo, then, `npm install` the Node dependencies. Then, make sure you've installed Xcode. If you haven't already, install [CocoaPods](https://cocoapods.org), and then run `pod install` in the `ios` directory. Now that you're mostly setup, you should be able to run `npm start` in one terminal, and then `npm run ios` in another. An iOS simulator should be launched containing the app. That concludes the first part.

Once you've done that, you can get to crafting a CarPlay integration. From what I see, the most straightforward package for doing so is [react-native-carplay](https://github.com/birkir/react-native-carplay). You should be able to set it up using the instructions that are provided in the repository and run it in the simulator.

Now comes the hard part. From what I can see, you don't have full control over the CarPlay UI, but have to rely on a set of [templates that are supplied by Apple](https://developer.apple.com/carplay/documentation/CarPlay-App-Programming-Guide.pdf). Good news: they look kinda good, and there's one specifically for a Now Playing screen (see PDF). Bad news: this template needs to be wrapped in `react-native-carplay`, and they [don't have a wrapper right now](https://github.com/birkir/react-native-carplay/issues/19). I don't think implementing it should be insanely hard, but you're going to have to dive in some Objective-C. 

If you get as far as fixing this, implementing all the data bindings should be a breeze, since I've built some hooks (`useCurrentTrack` and `useQueue`) to retrieve all the relevant data.

Feel free to fork the repo and create a draft Pull Request. More than happy to track how you're faring and help out if necessary. Let me know if anything's unclear!

**Comment by isvein** [comment]
Any news here? :-)

**Comment by leinelissen** [comment]
It appears some people have been successful with getting this set up for CarPlay (https://github.com/birkir/react-native-carplay/issues/175). I think this should be feasible for someone willing to take it up. I'm curious to look at it, but still don't have time right now.

**Comment by develvic** [comment]
Any updates here? :)
